lib/mail_room/connection.rb

```diff
@@ -0,0 +1,170 @@
+module MailRoom
+  class Connection
+    def initialize(mailbox)
+      @mailbox = mailbox
+
+      # log in and set the mailbox
+      reset
+      setup
+    end
+
+    def on_new_message(&block)
+      @new_message_handler = block
+    end
+
+    # is the connection logged in?
+    # @return [Boolean]
+    def logged_in?
+      @logged_in
+    end
+
+    # is the connection blocked idling?
+    # @return [Boolean]
+    def idling?
+      @idling
+    end
+
+    # is the imap connection closed?
+    # @return [Boolean]
+    def disconnected?
+      @imap.disconnected?
+    end
+
+    # is the connection ready to idle?
+    # @return [Boolean]
+    def ready_to_idle?
+      logged_in? && !idling?
+    end
+
+    def quit
+      stop_idling
+      reset
+    end
+
+    def wait
+      begin
+        # in case we missed any between idles
+        process_mailbox
```

Comment:
@godfat I'm processing the mailbox before and after idle. If idle raises, the second process won't happen so messages should remain untouched. If the message being fetched causes the Net::IMAP issue, it may be marked as SEEN, and would not be fetched again. I don't know GOOGLE's internal implementation of that behavior. It also depends on using UNSEEN as the search command like I do.

Debating if process_mailbox should happen outside of begin/rescue as per issue #72.